internal/analysis/pairwise.go

```diff
@@ -105,7 +105,11 @@ func PairwiseAnalysis(options AnalysisOptions) (map[model.PairwiseKey]*model.Pai
 	}
 
 	if len(output.WorkErrors) != 0 {
-		return nil, 0, fmt.Errorf("Failed to run pairwise analysis for '%d' submissions.", len(output.WorkErrors))
+		for pairwiseKey, err := range output.WorkErrors {
+			log.Error("Failed to run pairwise analysis.", err, log.NewAttr("pairwise-key", pairwiseKey.String()), log.NewAttr("job-id", output.ID))
```

Comment:
Logging the components here is a bit more tricky, since there could be multiple elements.
You may need to see if there are multiple keys, sort them, use the first key as the primary thing (course, assignment, user), and then put the second course in some alternate key (e.g., "alt-course"?).

Since this is complex, I would probably stash it away in a LogValue() method:
https://github.com/edulinq/autograder-server/blob/main/internal/log/attribute.go#L20